### PR TITLE
Feature/new active force

### PR DESCRIPTION
**Pull request by SchoeniPhlippsn**
## Description

Converting Active forces to the new version 3 API.
Additionally, I simplified the active force based on issue #633, except for keeping the original code of constraint forces. This has to be changed once the general `manifold` class is implemented.

The only remaining parameters are:
1. `filter`
2. `seed`
3. `rotation_diff`
4. `constraint` Has to be changed to `manifold` later and is set to `None` for now
The others have been removed. (`orientation_link` now always `true`)

The active force and torque are now given in the local particle frame and defined per particle type instead of per particle. 
  
## Motivation and Context
Resolves: #633 
Ports active forces to v3 API
## How Has This Been Tested?

I have added a pytest to check if active forces attach properly. I also checked for ballistic behaviour within Langevin dynamics 

## Change log

```
Simplified and moved Active forces to v3 API
```

## Checklist:

- [x] I have reviewed the [**Contributor Guidelines**](https://github.com/glotzerlab/hoomd-blue/blob/master/CONTRIBUTING.md).
- [x] I agree with the terms of the [**HOOMD-blue Contributor Agreement**](https://github.com/glotzerlab/hoomd-blue/blob/master/ContributorAgreement.md).
- [ ] My name is on the [list of contributors](https://github.com/glotzerlab/hoomd-blue/blob/master/sphinx-doc/credits.rst).

### CODE REVIEW

File: hoomd/md/ActiveForceComputeGPU.cu [hunk 1]
```diff
@@ -57,57 +56,21 @@ __global__ void gpu_compute_active_force_set_forces_kernel(const unsigned int gr
 
     unsigned int tag = d_groupTags[group_idx];
     unsigned int idx = d_rtag[tag];
```
Comment (joaander) [inline hoomd/md/ActiveForceComputeGPU.cu:58]: Remove `m_groupTags` and access the indices directly from the group via `m_group->getIndexArray()`.
Now that the directions are tied to orientation, the computation does not need the particle tag.

File: hoomd/md/ActiveForceComputeGPU.cu [hunk 2]
```diff
@@ -57,57 +56,21 @@ __global__ void gpu_compute_active_force_set_forces_kernel(const unsigned int gr
...
-        d_force[idx].z = fi.z;
-
-        vec3<Scalar> ti;
-        t = make_scalar3(d_t_actMag[tag] * d_t_actVec[tag].x,
-                        d_t_actMag[tag] * d_t_actVec[tag].y, d_t_actMag[tag] * d_t_actVec[tag].z);
-        ti = rotate(quati, vec3<Scalar>(t));
-        d_torque[idx].x = ti.x;
-        d_torque[idx].y = ti.y;
-        d_torque[idx].z = ti.z;
+    Scalar3 f = make_scalar3(d_f_actMag[type]*d_f_actVec[type].x, d_f_actMag[type]*d_f_actVec[type].y, d_f_actMag[type]*d_f_actVec[type].z);
```
Comment (joaander) [inline hoomd/md/ActiveForceComputeGPU.cu:61]: I don't expect this kernel to take very long to run as is, so this micro optimization isn't necessary. If you want to refactor the storage, we could reduce the number of memory transactions by packing the vector and magnitude into a single Scalar4 and reading it with ``__ldg``:
```
Scalar4 act = __ldg(d_f_act + type);
vec3<Scalar> f(act.x * act.w, act.y * act.w, act.z * act.w)
```

File: hoomd/md/force.py [hunk 3]
```diff
@@ -271,73 +280,48 @@ class active(_Force):
```
Comment (joaander) [inline hoomd/md/force.py:280]: I think the note about the lack of MPI execution support should be removed. The new code will work in MPI.

File: hoomd/md/force.py [hunk 4]
```diff
@@ -238,24 +253,18 @@ def update_coeffs(self):
...
-        group (:py:mod:`hoomd.group`): Group for which the force will be set
-        orientation_link (bool): if True then forces and torques are applied in the particle's reference frame. If false, then the box
-         reference frame is used. Only relevant for non-point-like anisotropic particles.
-        orientation_reverse_link (bool): When True, the particle's orientation is set to match the active force vector. Useful for
-         for using a particle's orientation to log the active force vector. Not recommended for anisotropic particles. Quaternion rotation
-         assumes base vector of (0,0,1).
-        rotation_diff (float): rotational diffusion constant, :math:`D_r`, for all particles in the group.
         constraint (:py:class:`hoomd.md.update.constraint_ellipsoid`) specifies a constraint surface, to which particles are confined,
           such as update.constraint_ellipsoid.
+        rotation_diff (float): rotational diffusion constant, :math:`D_r`, for all particles in the group.
```
Comment (joaander) [inline hoomd/md/force.py:266]: Add an `Attributes:` section and document `filter`, `seed`, `rotation_diff`, `constraint`, `active_force` and `active_torque`)

**Comment by SchoeniPhlippsn** [comment]
Hi Josh,
Thanks for your comments. I implemented all of your recommended changes yesterday. I hope all the issues have been resolved now.

File: hoomd/md/ActiveForceComputeGPU.cc [hunk 5]
```diff
@@ -27,94 +27,72 @@ using namespace std;
...
-    ArrayHandle<Scalar> d_t_actMag(m_t_activeMag, access_location::device, access_mode::read);
+    ArrayHandle<Scalar4> d_t_actVec(m_t_activeVec, access_location::device, access_mode::read);
     ArrayHandle<Scalar4> d_torque(m_torque, access_location::device, access_mode::overwrite);
 
-    ArrayHandle<Scalar4> d_orientation(m_pdata->getOrientationArray(), access_location::device, access_mode::readwrite);
+    ArrayHandle<Scalar4> d_pos(m_pdata -> getPositions(), access_location::device, access_mode::read);
+    ArrayHandle<Scalar4> d_orientation(m_pdata->getOrientationArray(), access_location::device, access_mode::read);
     ArrayHandle<unsigned int> d_rtag(m_pdata->getRTags(), access_location::device, access_mode::read);
-    ArrayHandle<unsigned int> d_groupTags(m_groupTags, access_location::device, access_mode::read);
+    ArrayHandle<unsigned int> d_groupTags(m_group->getIndexArray(), access_location::device, access_mode::read);
```
Comment (joaander) [inline hoomd/md/ActiveForceComputeGPU.cc:86]: We should label this as `d_index_array` so we don't confuse them with tags.

File: hoomd/md/ActiveForceCompute.cc [hunk 6]
```diff
@@ -173,81 +216,64 @@ void ActiveForceCompute::setForces()
         {
         unsigned int tag = m_group->getMemberTag(i);
         unsigned int idx = h_rtag.data[tag];
```
Comment (joaander) [inline hoomd/md/ActiveForceCompute.cc:218]: No need to use tag indirection, access the index directly.
`unsigned int idx = m_group->getMemberIndex(group_idx);`

File: hoomd/md/ActiveForceCompute.cc [hunk 7]
```diff
@@ -332,34 +339,62 @@ void ActiveForceCompute::setConstraint()
...
     ArrayHandle<unsigned int> h_rtag(m_pdata->getRTags(), access_location::host, access_mode::read);
+
+    assert(h_f_actVec.data != NULL);
     assert(h_pos.data != NULL);
+    assert(h_orientation.data != NULL);
 
     for (unsigned int i = 0; i < m_group->getNumMembers(); i++)
         {
         unsigned int tag = m_group->getMemberTag(i);
         unsigned int idx = h_rtag.data[tag];
```
Comment (joaander) [inline hoomd/md/ActiveForceCompute.cc:354]: Use `getMemberIndex` here as well.

File: hoomd/md/ActiveForceCompute.cc [hunk 8]
```diff
@@ -173,81 +216,64 @@ void ActiveForceCompute::setForces()
...
```
Comment (joaander) [inline hoomd/md/ActiveForceCompute.cc:256]: Use `getMemberIndex` here as well.

File: hoomd/md/ActiveForceComputeGPU.cc [hunk 9]
```diff
@@ -145,13 +120,11 @@ void ActiveForceComputeGPU::setForces()
...
-    ArrayHandle<Scalar3> d_f_actVec(m_f_activeVec, access_location::device, access_mode::readwrite);
-    ArrayHandle<Scalar3> d_t_actVec(m_t_activeVec, access_location::device, access_mode::readwrite);
+    ArrayHandle<Scalar4> d_f_actVec(m_f_activeVec, access_location::device, access_mode::readwrite);
     ArrayHandle<Scalar4> d_pos(m_pdata -> getPositions(), access_location::device, access_mode::read);
-    ArrayHandle<Scalar4> d_force(m_force, access_location::device, access_mode::overwrite);
-    ArrayHandle<Scalar4> d_torque(m_torque, access_location::device, access_mode::overwrite);
+    ArrayHandle<Scalar4> d_orientation(m_pdata->getOrientationArray(), access_location::device, access_mode::readwrite);
     ArrayHandle<unsigned int> d_rtag(m_pdata->getRTags(), access_location::device, access_mode::read);
-    ArrayHandle<unsigned int> d_groupTags(m_groupTags, access_location::device, access_mode::read);
+    ArrayHandle<unsigned int> d_groupTags(m_group->getIndexArray(), access_location::device, access_mode::read);
```
Comment (joaander) [inline hoomd/md/ActiveForceComputeGPU.cc:127]: Will need to use `d_index_array` everywhere that `d_groupTags` is currently.

File: hoomd/md/ActiveForceComputeGPU.cu [hunk 10]
```diff
@@ -38,76 +36,42 @@ __global__ void gpu_compute_active_force_set_forces_kernel(const unsigned int gr
...
                                                     const unsigned int N)
     {
     unsigned int group_idx = blockIdx.x * blockDim.x + threadIdx.x;
     if (group_idx >= group_size)
         return;
 
-    unsigned int tag = d_groupTags[group_idx];
-    unsigned int idx = d_rtag[tag];
+    unsigned int tag = __ldg(d_groupTags + group_idx);
+    unsigned int idx = __ldg(d_rtag + tag);
```
Comment (joaander) [inline hoomd/md/ActiveForceComputeGPU.cu:54]:
```suggestion
    unsigned int idx = d_index_array[group_idx];
```

File: hoomd/md/ActiveForceComputeGPU.cu [hunk 11]
```diff
@@ -126,53 +89,65 @@ __global__ void gpu_compute_active_force_set_constraints_kernel(const unsigned i
...
+                                                   const Scalar rz)
     {
     unsigned int group_idx = blockIdx.x * blockDim.x + threadIdx.x;
     if (group_idx >= group_size)
         return;
 
-    unsigned int tag = d_groupTags[group_idx];
-    unsigned int idx = d_rtag[tag];
+    unsigned int tag = __ldg(d_groupTags + group_idx);
+    unsigned int idx = __ldg(d_rtag + tag);
```
Comment (joaander) [inline hoomd/md/ActiveForceComputeGPU.cu:104]:
```suggestion
    unsigned int idx = d_index_array[group_idx];
```

File: hoomd/md/ActiveForceComputeGPU.cu [hunk 12]
```diff
@@ -210,80 +184,83 @@ __global__ void gpu_compute_active_force_rotational_diffusion_kernel(const unsig
...
```
Comment (joaander) [inline hoomd/md/ActiveForceComputeGPU.cu:188]:
```suggestion
    unsigned int idx = d_index_array[group_idx];
```

**Comment by SchoeniPhlippsn** [comment]
> Great, thanks!
> 
> I merged in a set of changes that fixes the CI tests that were failing due to errors on the parent branch. During the merge, I fixed compile errors on the GPU. Could you check or test the new GPU code and make sure it works correctly?

I checked the GPU code and everything works correctly! In that regard, no changes are needed.

**Comment by joaander** [comment]
Thanks!

**Pull request merged**